### PR TITLE
ComponentManager Rework

### PR DESCRIPTION
**Pull request by Tyler-IN**
Requires space-wizards/RobustToolbox#1143

fix benchmark

improve some test diagnostics

fix some bug where order of shutdown of AiControllerComponent mattered or Processor was never initialized

### CODE REVIEW

File: Content.Benchmarks/ComponentManagerGetAllComponents.cs [hunk 1]
```diff
@@ -65,6 +88,16 @@ public int Run()
             return count;
         }
 
+        [Benchmark]
+        public int Noop()
```
Comment (Tyler-IN) [inline Content.Benchmarks/ComponentManagerGetAllComponents.cs:92]: Suggest a better name please.

File: Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs [hunk 2]
```diff
@@ -60,7 +60,7 @@ public override void ExposeData(ObjectSerializer serializer)
             base.ExposeData(serializer);
 
             serializer.DataField(ref StorageCapacityMax, "Capacity", 10000);
-            serializer.DataField(ref StorageUsed, "used", 0);
+            //serializer.DataField(ref StorageUsed, "used", 0);
```
Comment (Tyler-IN) [inline Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs:63]: Maybe should remove both of these, idk.

File: Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs [hunk 3]
```diff
@@ -348,7 +348,7 @@ private void _ensureInitialCalculated()
 
             foreach (var entity in storage.ContainedEntities)
             {
-                var item = entity.GetComponent<ItemComponent>();
+                var item = entity.GetComponent<StoreableComponent>();
```
Comment (Tyler-IN) [inline Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs:351]: Follows pattern of other ObjectSize summation.

File: Content.Server/GameObjects/Components/Movement/AiControllerComponent.cs [hunk 4]
```diff
@@ -61,7 +61,7 @@ public override void ExposeData(ObjectSerializer serializer)
         protected override void Shutdown()
         {
             base.Shutdown();
-            Processor.Shutdown();
+            Processor?.Shutdown();
```
Comment (Tyler-IN) [inline Content.Server/GameObjects/Components/Movement/AiControllerComponent.cs:64]: Can anyone tell me why this is necessary now? It's _fine_ but I don't know why.

File: Content.Benchmarks/Program.cs [hunk 5]
```diff
@@ -6,7 +6,8 @@ internal static class Program
     {
         public static void Main(string[] args)
         {
-            BenchmarkRunner.Run<StereoToMonoBenchmark>();
+            BenchmarkRunner.Run<ComponentManagerGetAllComponents>();
+            //ComponentManagerGetAllComponents.TestRun();
```
Comment (Tyler-IN) [inline Content.Benchmarks/Program.cs:10]: Revert?
```suggestion
           BenchmarkRunner.Run<StereoToMonoBenchmark>();
```

File: Content.IntegrationTests/Tests/EntityTest.cs [hunk 6]
```diff
@@ -69,7 +69,8 @@ public async Task Test()
                         catch (Exception e)
                         {
                             Logger.LogS(LogLevel.Error, "EntityTest", "Entity '" + prototype.ID + "' threw: " + e.Message);
-                            Assert.Fail();
+                            //Assert.Fail();
+                            throw;
```
Comment (Tyler-IN) [inline Content.IntegrationTests/Tests/EntityTest.cs:73]:
```suggestion
                            throw;
```

File: Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs (hunk 2 above)
Comment (Tyler-IN) [inline Content.Server/GameObjects/Components/Items/Storage/ServerStorageComponent.cs:63]: Maybe should remove both of these, idk.
```suggestion
```

File: Content.Benchmarks/ComponentManagerGetAllComponents.cs [hunk 7]
```diff
@@ -14,7 +18,17 @@ public class ComponentManagerGetAllComponents
...
+
+        public static void TestRun()
+        {
+            var x = new ComponentManagerGetAllComponents
+            {
+                N = 500
+            };
+            x.Setup();
+            x.Run();
+        }
```
Comment (Tyler-IN) [inline Content.Benchmarks/ComponentManagerGetAllComponents.cs:31]: For debugging w/o running the entire benchmark.

**Comment by Tyler-IN** [comment]
Mosts tests complete just as fast or faster. Shaved some time off of larger tests.

Benchmark run for GetAllComponents:
old ComponentManager: 57,611.27 ns,
new ComponentManager: 82.34 ns

File: Content.Server/GameObjects/Components/Movement/AiControllerComponent.cs (hunk 4 above)
Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Movement/AiControllerComponent.cs:64]: The AiSystem initializes the processor so if it's not getting run in the tests that could cause it. I'd need to look further into it.

**Pull request merged**